runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -12358,6 +12358,936 @@ static TR::Register *inlineIntrinsicIndexOf(TR::Node *node, TR::CodeGenerator *c
+      {
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, arrayAddressReg, arrayObjReg,
+                                     TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, subStringAddressReg, subStringObjReg,
+                                     TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+      }
+
+   /* resultReg currently holds the starting offset. currentAddressReg will hold the address of the first byte of the main string to look at. */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, currentAddressReg, arrayAddressReg, resultReg);
+
+   /*
+    * This is a early check to see if there is an immediate match for the substring in the main string.
+    * The first character of the substring is compared against the first (after adjusted for offset) character of the main string.
+    * If it matches, jump to foundEarlyLabel to try and match the rest of the substring. If it doesn't match, jump back to notFoundEarlyLabel.
+    */
+   generateTrg1MemInstruction(cg, isLatin1 ? TR::InstOpCode::lbz : TR::InstOpCode::lhz, node, tempReg, TR::MemoryReference::createWithDisplacement(cg, currentAddressReg, 0, isLatin1 ? 1 : 2));
+   generateTrg1MemInstruction(cg, isLatin1 ? TR::InstOpCode::lbz : TR::InstOpCode::lhz, node, zeroReg, TR::MemoryReference::createWithDisplacement(cg, subStringAddressReg, 0, isLatin1 ? 1 : 2));
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, cr0Reg, tempReg, zeroReg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, foundEarlyLabel, cr0Reg);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, notFoundEarlyLabel);
+
+   /*
+    * IMPORTANT: The upper 32 bits of a 64-bit register containing an int are undefined. Since the
+    * indices are being passed in as ints, we must ensure that their upper 32 bits are not garbage.
+    */
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::extsw, node, endAddressReg, arrayLenReg);
+
+   /* Under UTF16, the indices are doubled since offset and size in bytes are needed later. */
+   if (!isLatin1)
+      {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, endAddressReg, endAddressReg, endAddressReg);
+      }
+
+   /*
+    * It sometimes isn't necessary to match the first character of the substring against the last characters in the main string
+    * because there aren't enough characters left in the main string to hold the rest of the substring.
+    * A check for this case is performed here and goes to notFoundLabel to return -1 if there aren't enough characters left.
+    */
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, endAddressReg, subStringLenReg, endAddressReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp8, node, cr0Reg, resultReg, endAddressReg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bge, node, notFoundLabel, cr0Reg);
```

Comment:
compare-equal is included in this branch ... due to NotFoundEarly and before offset-reg (resultReg) is adjusted. walking on a pretty thin line ... way of programming.

Reply:
I also thought this was very awkward. But, doing it this way let me avoid incrementing `endAddressReg` if the branch is taken. I made sure to explicitly test this area to make sure the math is right.